### PR TITLE
use different Visual Studio bootstrappers for apex/e2e tests and optprof pipeline

### PR DESCRIPTION
**Pull request by kartheekp-ms**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/1889

Regression? Last working version:

## Description

Currently, CI produces a single VS bootstrapper that is used for running VS apex/e2e tests & OptProf pipeline on DartLab. In very rare cases we may have to run VS apex/e2e tests on a different channel such as `intpreview` or `release` instead of `int.main`. The reason for this is any regression in `int.main` branch causing VS apex/e2e tests to fail will block team members from merging their pull requests. Refer to https://github.com/NuGet/Home/issues/12104 and for one example that impacted team velocity during Hackathon 2022 week. Official builds failed last week because of https://developercommunity.visualstudio.com/t/Visual-Studio-freezes-and-needs-to-be-ki/10182087 issue.

This PR addresses @zivkan comment in https://github.com/NuGet/NuGet.Client/pull/4821#discussion_r980707961 to `Maybe it's worthwhile having the OptProf pipeline build its own bootstrapper, that way the build pipeline only builds one (used for E2E and Apex), and we don't need to worry about OptProf re-using it.`

The way I tried to implement this change is by default we use `VSTargetChannel` from config.props for running Apex/E2E tests. In an event where a regression is blocking our builds then we can override the `VS channel` on which we run the tests by overriding `VSTargetChannelForTests` pipeline variable. I added this variable to both PR and official builds.

I triggered [official build](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6897274&view=results) and [optprof pipeline](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6898381&view=results) for this private build and both of them succeeded. 

There are few improvements/clean-ups that can be made to this PR, but I would like to merge it as soon as possible to unblock official CI builds and do the clean-ups in the follow-up PR.

- Do we need to add `VSTargetChannelForTests` in config.props just like `VSSTargetChannel` property?https://github.com/NuGet/NuGet.Client/blob/6a3a3400c4778de9c7f55c97fc21464bbc9b579c/build/config.props#L29
- Can we publish the `MicroBuildOutputs` in a single task instead of two separate tasks introduced in this PR?
 https://github.com/NuGet/NuGet.Client/blob/dev/eng/pipelines/templates/Build_and_UnitTest.yml#L334-L340

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] N/A 

- **Documentation**
  
  - [x] N/A

### CODE REVIEW

File: eng/pipelines/templates/Build_and_UnitTest.yml
Comment (nkolev92) [inline eng/pipelines/templates/Build_and_UnitTest.yml]: Given that these artifacts are created in the same directory, do we have any overlap concerns? Or does the above task create the json file with a different name.

File: eng/pipelines/templates/Build_and_UnitTest.yml
Comment (zivkan) [inline eng/pipelines/templates/Build_and_UnitTest.yml]: Do we have any guarantees that the second bootstrapper will always be index `[1]`?  Perhaps Nikolche's suggestion of building the bootstrapper in a second directory will remove the risk.

File: eng/pipelines/templates/Build_and_UnitTest.yml
Comment (zivkan) [inline eng/pipelines/templates/Build_and_UnitTest.yml]: The size of the bootstrapper isn't huge, but it is megs. I'm thinking about the amount of storage that we'll use for all our retained builds if we start doubling the number of bootstrappers we have.  Although honestly, we should probably move the bootstrapper out of the VS15 directory and put it somewhere that isn't retained.

What do you think of another idea, where the PR/official pipelines still only builds one bootstrapper, but the OptProf pipeline creates its own bootstrapper? OptProf pipeline runs aren't retained for long, so we'll never need to worry about long term storage costs.

File: eng/pipelines/templates/Build_and_UnitTest.yml
Comment (kartheekp-ms) [reply eng/pipelines/templates/Build_and_UnitTest.yml]: Great question Nikolche. I wanted to address this in the PR description but missed it. In the PR build and official build, I triggered for this change, I noticed that a single `bootstrapper.json` file got generated with two entries. First entry pointing to `bootstrapper for optprof` because it was generated first in the pipeline and the second entry pointing to `bootstrapper for tests` because it was generated later in the pipeline.

For example, private build for this PR produced following `bootstrapper.json` file

```json
[
    {
        "overlaidChannelManifestUrl":  "REMOVED",
        "VSBranch":  "main",
        "VSManifestName":  "VisualStudio.17.int.main",
        "VSBuildVersion":  "17.5.33031.37",
        "BuildDrop":  "REMOVED/0110c214-cfdb-b409-3101-6609d74dc856",
        "InputManifests":  "D:\\a\\_work\\1\\s\\artifacts\\VS15\\Microsoft.VisualStudio.NuGet.Core.vsman",
        "bootstrapperUrl":  "REMOVED/fc5794c1-45c5-487f-9254-27215b06a90e/vs_enterprise.exe"
    },
    {
        "overlaidChannelManifestUrl":  "REMOVED",
        "VSBranch":  "d17.4",
        "VSManifestName":  "VisualStudio.17.int.d17.4",
        "VSBuildVersion":  "17.4.33027.239",
        "BuildDrop":  "REMOVED/1b6eebc8-6b11-52dd-106f-345caa00c389",
        "InputManifests":  "D:\\a\\_work\\1\\s\\artifacts\\VS15\\Microsoft.VisualStudio.NuGet.Core.vsman",
        "bootstrapperUrl":  "REMOVED/c9353840-ed32-4749-abf9-b129739292ba/vs_enterprise.exe"
    }
]
```

Another notable difference here is that the GUID in `bootstrapperUrl` in the above json file is different for `int.main` channel and `.int.d17.4` channel.

The artifacts uploaded to the build pipeline also confirms the same.

image

The subfolder with guid `c9353840-ed32-4749-abf9-b129739292ba` is used for apex/e2e tests and folder name whereas the subfolder with guid `fc5794c1-45c5-487f-9254-27215b06a90e` is used for optprof pipeline.

File: eng/pipelines/templates/Build_and_UnitTest.yml
Comment (kartheekp-ms) [reply eng/pipelines/templates/Build_and_UnitTest.yml]: I didn't think about the storage cost at all. Thank you, Andy, for reminding me.

> What do you think of another idea, where the PR/official pipelines still only builds one bootstrapper, but the OptProf pipeline creates its own bootstrapper? 

I thought about that, but I wanted to get a hacky solution working to unblock official builds. In the latest commit, I addressed your comment please review and let me know the feedback. [OptProf pipeline](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6905358&view=results) and [official build](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6904450&view=results) succeeded with the latest changes.

File: eng/pipelines/templates/Build_and_UnitTest.yml
Comment (kartheekp-ms) [reply eng/pipelines/templates/Build_and_UnitTest.yml]: > Do we have any guarantees that the second bootstrapper will always be index [1]? Perhaps Nikolche's suggestion of building the bootstrapper in a second directory will remove the risk.

Thank you @zivkan and @nkolev92 for the review comments because they pushed me to get the cleaner approach working. When I created this PR, I know the proposed solution is hacky to unblock official builds. That's why I mentioned `There are few improvements/clean-ups that can be made to this PR, but I would like to merge it as soon as possible to unblock official CI builds and do the clean-ups in the follow-up PR.` in the PR description. Please review the latest commit and let me know your feedback. 

 [OptProf pipeline](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6905358&view=results) and [official build](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6904450&view=results) succeeded with the latest changes.

File: scripts/cibuild/ConfigureVstsBuild.ps1
Comment (kartheekp-ms) [inline scripts/cibuild/ConfigureVstsBuild.ps1:177]: `buildinfo.json` now contains 2 additional properties `$VsTargetChannel` & `$VsTargetMajorVersion` which are used for generating VS bootstrapper in OptProf pipeline.

File: eng/pipelines/templates/Initialize_Build.yml
Comment (kartheekp-ms) [inline eng/pipelines/templates/Initialize_Build.yml:33]: If `VsTargetChannelOverrideForTests` pipeline variable is configured, then Apex/E2E tests will run on that channel instead of default `int.main` channel.

File: eng/pipelines/optprof.yml
Comment (zivkan) [inline eng/pipelines/optprof.yml]: This isn't needed. Honestly, we shouldn't be publishing the bootstrapper as build artifacts in our build pipeline, but there's only so much time in a day, and far too many things to work on.

File: eng/pipelines/optprof.yml
Comment (zivkan) [inline eng/pipelines/optprof.yml:174]: This might clash with an official build. Well, if the build numbers used the same format, which I don't think is, so perhaps it's unintentially unique.  Fortunately, drops are write-once, so if this pipeline tried to overwrite an official build drop, this pipeline would fail.

```suggestion
        buildNumber: 'OptProf/$(System.TeamProject)/$(Build.Repository.Name)/$(Build.SourceBranchName)/$(Build.BuildNumber)'
```

However, I'm not sure what other consequences there are. Like, how do we tell the bootstrapper task that we want to use a different prefix?

**Review by zivkan — COMMENTED** [review]
I'm worried about the theoretical possibility of the optprof pipeline trying to overwrite the offivial build drop. Could you see if it's possible to get the bootstrapper task to use a different URL prefix in the bootstrapperinfo.json file?

File: eng/pipelines/optprof.yml
Comment (kartheekp-ms) [reply eng/pipelines/optprof.yml]: Thanks Andy removed it from the pipeline.

File: eng/pipelines/optprof.yml
Comment (kartheekp-ms) [reply eng/pipelines/optprof.yml:174]: `buildNumber` will not clash with the official build because the last parameter `$(Build.BuildNumber)` is different between official build and optprof pipeline. As you know for official build, we generate build number from config.props https://github.com/NuGet/NuGet.Client/blob/dev/build/config.props#L15-L18. Whereas for [optprof pipeline](https://devdiv.visualstudio.com/DevDiv/_build?definitionId=16906&_a=summary) the build number is automatically generated by Azure DevOps.

Comment (kartheekp-ms) [reply eng/pipelines/optprof.yml:174]: > Whereas for [optprof pipeline](https://devdiv.visualstudio.com/DevDiv/_build?definitionId=16906&_a=summary) the build number is automatically generated by Azure DevOps.

Build number is actually set by a task in the OptProf pipeline. https://github.com/NuGet/NuGet.Client/blob/dev/eng/pipelines/optprof.yml#L73-L78. I added `Build.BuildNumber` value to the build number just to be able to run OptProf pipeline for the same official build. It was not the case earlier because `Upload VSTS drop task` failed with `Drop Already Exists` exception in my testing.

> However, I'm not sure what other consequences there are. 

I tried changing the prefix to `OptProf` but the [`Upload VSTS drop` task failed](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6908721&view=logs&j=87ee994e-8d25-5c3a-5836-c28a1c54ff14&t=12c880cb-a8ee-5c18-26cb-685efba1b642)  with `##[error]Cannot find path '%temp%\OptProf\DevDiv\NuGet-NuGet.Client-Trusted\dev-kartheekp-ms-differentvsbootstrappers-useOptProfForBuildNumber\dev-kartheekp-ms-differentvsbootstrappers-useOptProfForBuildNumber_7de18ecea5607b9d34cef4d4dd4f7074cbc8d1ca' because it does not exist.` error (see `OptProf` in the file path printed in the error message.)

I triggered OptProf pipeline on the latest commit and it [passed](https://devdiv.visualstudio.com/DevDiv/_build/results?buildId=6910076&view=results). This build has `Build.BuildNumber` appended to build name as mentioned above so that it creates unique drops for every build even when we have to run optprof pipeline multiple times for the same official build.

I think at this point it is safe to merge this PR because build number of OptProf pipeline will not clash with Official build.

**Review by nkolev92 — APPROVED** [review]
This makes sense to me. 
I'd wait for Andy's approval before merging :)

File: eng/pipelines/optprof.yml
Comment (zivkan) [reply eng/pipelines/optprof.yml:174]: > I tried changing the prefix to OptProf but the Upload VSTS drop task failed with

If you read the log, it says that the drop finalization was complete, and further up there's a link to "view the drop from a browser", where we can see all the files are there.

I'm guessing that the error happened when the task is cleaning up after itself, and you'll notice that the path it listed exceeds Windows' MAX_PATH, due to the extremely long branch name and build number used. My bet is that the machine either doesn't have long path enabled, or the process that the task is running doesn't have the right win32 manifest. Both are necessary for long paths to work, it's not enough to only enable the registry key ([for example](https://github.com/NuGet/NuGet.Client/pull/4830)).

> This build has Build.BuildNumber appended to build name

sorry to be pedantic, just trying to make sure you understand how the build works in case it helps reduce confusion in the future. There is no variable "build name". The script further up overwrites BuildNumber, at which point the previous BuildNumber is no longer available. Your change was to append the auto-generated BuildNumber in the new value for BuildNumber.

I find it really unfortunate that the MicroBuildBuildVSBootstrapper task doesn't appear to have any parameters to change the URL pattern it uses. This pipeline is a bit "special", since this build itself is not the important thing, the "component under test" is. As a consequence it would be ideal for us to keep the BuildNumber "clean" for when viewing the list of builds the pipeline has run, but we need to use a different, unique number just for the drop URL.

**Review by zivkan — APPROVED** [review]
I'm a bit hesitant to use this on the official builds, since it will cause release-* branches to also use the override, at which time they might fail because they'll be running against versions of VS that they're really not designed for.  But maybe I'm making perfect the enemy of good. We're much more impacted by VS main branch breaking NuGet's dev branch.

**Comment by kartheekp-ms** [comment]
> I'm a bit hesitant to use this on the official builds, since it will cause release-* branches to also use the override, at which time they might fail because they'll be running against versions of VS that they're really not designed for. But maybe I'm making perfect the enemy of good. We're much more impacted by VS main branch breaking NuGet's dev branch.

I will address this in a follow-up PR shortly. Thank you for such detailed comments. Very helpful.

**Pull request merged**
